### PR TITLE
Support to pass the image name to the remote mount slice

### PR DESCRIPTION
**Pull request by ChengyuZhu6**
In the scenario of confidential containers, the image needs to be downloaded inside the vm. So an external method is needed to pass the name of the image to the vm. Therefore, we hope the snapshotter can pass the image name information through the mount slice.

Fixes: #516

### CODE REVIEW

File: snapshot/snapshot.go [hunk 1]
```diff
@@ -716,7 +718,25 @@ func overlayMount(options []string) []mount.Mount {
 		},
 	}
 }
+func (o *snapshotter) passImageNameMounts(ctx context.Context, parent string) ([]mount.Mount, error) {
+	_, pInfo, _, err := snapshot.GetSnapshotInfo(ctx, o.ms, parent)
+	if err != nil {
+		return nil, errors.Wrapf(err, "passImageNameMounts get snapshot %q info failed", parent)
+	}
+	passImageOption := fmt.Sprintf("imageurl=%s", pInfo.Labels["containerd.io/snapshot/cri.image-ref"])
```
Comment (changweige) [inline snapshot/snapshot.go:726]: I suppose it's possible that the key `containerd.io/snapshot/cri.image-ref` does not exist in snapshot's labels. We should handle such case

File: snapshot/snapshot.go [hunk 2]
```diff
@@ -716,7 +718,25 @@ func overlayMount(options []string) []mount.Mount {
...
 
+	var overlayOptions []string
+	overlayOptions = append(overlayOptions, passImageOption)
+
+	return []mount.Mount{
+		{
+			Type:    "fuse.nydus-overlayfs",
```
Comment (changweige) [inline snapshot/snapshot.go:733]: The filesystem or mount type `fuse.nydus-overlayfs` should only appear when nydus-snapshotter has enable_nydus_overlayfs enabled in its toml config.
Moreover, we should follow the what we do in `remoteMountWithExtraOptions` like encoding the mount options.

```
	// base64 to filter easily in `nydus-overlayfs`
	opt := fmt.Sprintf("extraoption=%s", base64.StdEncoding.EncodeToString(no))
	overlayOptions = append(overlayOptions, opt)
```

Comment (ChengyuZhu6) [reply snapshot/snapshot.go:739]: Thanks, @changweige . I have fixed it.

File: snapshot/snapshot.go (hunk 1 above)
Comment (ChengyuZhu6) [reply snapshot/snapshot.go]: Thanks for review. I have fixed it.

File: snapshot/snapshot.go [hunk 3]
```diff
@@ -716,7 +718,31 @@ func overlayMount(options []string) []mount.Mount {
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "passImageNameMounts get snapshot %q info failed", parent)
+	}
+	imageURL, ok := pInfo.Labels["containerd.io/snapshot/cri.image-ref"]
 
+	if !ok {
+		return nil, errors.Errorf("passImageNameMounts failed to find image ref of snapshot %s, labels %v",
+			parent, pInfo.Labels)
+	}
+	passImageOption := fmt.Sprintf("imageurl=%s", base64.StdEncoding.EncodeToString([]byte(imageURL)))
```
Comment (imeoer) [inline snapshot/snapshot.go:732]: How about follow https://github.com/containerd/containerd/blob/fb786b2f17b2c7d7ea5d9c9009d980bc03379952/pkg/snapshotters/annotations.go#L34, name it `image-ref` ?

File: snapshot/snapshot.go [hunk 4]
```diff
@@ -350,7 +350,9 @@ func (o *snapshotter) Mounts(ctx context.Context, key string) ([]mount.Mount, er
 	if err != nil {
 		return nil, errors.Wrapf(err, "get snapshot %s", key)
 	}
-
+	if o.enableNydusOverlayFS && config.GetPassImageURLEnabled() {
+		return o.passImageNameMounts(ctx, info.Parent)
```
Comment (imeoer) [inline snapshot/snapshot.go:354]: Can we move it into `remoteMountWithExtraOptions`?

File: config/config.go [hunk 5]
```diff
@@ -109,6 +109,7 @@ const (
 type Experimental struct {
 	EnableStargz         bool `toml:"enable_stargz"`
 	EnableReferrerDetect bool `toml:"enable_referrer_detect"`
+	EnablePassImageURL   bool `toml:"enable_pass_image_url"`
```
Comment (imeoer) [inline config/config.go:112]: Is it necessary to add the option?

**Pull request closed without merging**